### PR TITLE
Make iOS keyboard robust to selection of -1

### PR DESCRIPTION
**Pull request by abarth**
The Dart code uses -1 to signal that there is no current selection. If
we get text input in this state, we need to do something reasonable.
This patch teach us to treat that case as if the selection was zero.

This patch also improves how we convert from NSString to string16 and
makes it so we don't leak keyboard subviews.

Fixes https://github.com/flutter/flutter/issues/1878

### CODE REVIEW

**Comment by abarth** [comment]
@chinmaygarde

File: sky/services/editing/ios/keyboard_impl.mm [hunk 1]
```diff
@@ -57,13 +58,15 @@ - (void)show {
...
-  [[UIApplication sharedApplication].keyWindow addSubview:self];
+  if (!self.superview)
+    [[UIApplication sharedApplication].keyWindow addSubview:self];
   [self becomeFirstResponder];
 }
 
 - (void)hide {
   [self resignFirstResponder];
-  [self removeFromSuperview];
+  if (self.superview)
```
Comment (chinmaygarde) [inline sky/services/editing/ios/keyboard_impl.mm]: This check is redundant. Ditto for the `addSubview:` call above. Adding a view as a subview removes it from its previous parent.

**Comment by chinmaygarde** [comment]
lgtm

File: sky/services/editing/ios/keyboard_impl.mm [hunk 2]
```diff
@@ -134,17 +137,18 @@ - (void)deleteBackward {
     : binding_(this, request.Pass()), client_([[KeyboardClient alloc] init]) {}
 
 KeyboardImpl::~KeyboardImpl() {
+  [client_ hide];
```
Comment (abarth) [inline sky/services/editing/ios/keyboard_impl.mm]: Is this call necessary?  Is releasing the view sufficient to remove it from the parent?

Comment (chinmaygarde) [inline sky/services/editing/ios/keyboard_impl.mm]: The parent holds a strong reference to its view. So releasing the view will not remove it from its parent. The parent will own the remaining reference.

Comment (abarth) [inline sky/services/editing/ios/keyboard_impl.mm]: Ok.  I'll leave this one in then.

**Pull request merged**